Introducing Management Server

zalokhan opened the pull request:
### Management Server
Management server takes the serverContext and the layoutServer as parameters.
The layoutServer is used only initially to get the bootstrapping layout.

The server spawns a thread to run the fault detection as an async task.
(As pointed out by @Maithem this can be inefficient as there can be several threads running if we have 1000s of corfu servers on one node. So let me know if there is a better way to do this.)

The getRuntime method in the Management Server returns a connected instance of the corfuRuntime. It returns null if the server has been signaled to shutdown.

### Minor change in CorfuRuntime
I have added a field isShutdown in the corfu runtime to prevent it from going into an endless loop situation. (Especially when running tests.)
The invalidate corfuRuntime causes it to get stuck in the loop trying to fetch a layout even if all the layout servers have gone down.
Thus I've added the flag to break out of this.
This flag is checked only when the thread is interrupted. Else we avoid repetitive checks.

### Addition in CorfuServer
Initializing and starting the Management Server.

Maithem commented on src/main/java/org/corfudb/infrastructure/CorfuServer.java:
```diff
@@ -54,6 +54,9 @@
     @Getter
```
You need a git commit message.

zalokhan replied:
Yes I'll add this to the description.

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,236 @@
+     * Flag to notify management server of shutdown
+     */
+    @Getter
+    private boolean isShutdown = false;
```
this should be a volatile, it can be accessed by two different threads.

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,236 @@
+import java.util.concurrent.RejectedExecutionException;
+
+/**
+ * Instantiates and performs failure detection adn handling asynchronously.
```
Can you add a proper class message about how the failure detector works?

Also, "adn" -> "and"

zalokhan replied:
Check the latest commit.
Ive added a little more description and corrected the typo

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,236 @@
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.RejectedExecutionException;
```
Please use the import ordering convention here 
http://stackoverflow.com/questions/7735671/order-of-import-statements-in-java

zalokhan replied:
I'll group the imports.

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,236 @@
+    private final ExecutorService schedulerService;
+
+
+    public ManagementServer(ServerContext serverContext, LayoutServer layoutServer) {
```
Hmm, why do you need to pass this server?
You already have the runtime, you can get the runtime via
runtime.getLayoutView().getCurrentLayout();

zalokhan replied:
So as I have mentioned this in the above description, the layout server is required for initial bootstrapping. The management server needs to know which layout servers should it connect the runtime to. Once the runtime is connected, its no longer required.
Let me know if there is a workaround for this.

(I wanted to avoid giving the user another cmdlet to bootstrap the management server.)

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,236 @@
+            try {
+                Thread.sleep(1000);
+                currentLayout = layoutServer.getCurrentLayout();
+            } catch (InterruptedException err) {
```
Include err in the message.

zalokhan replied:
Done.

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,236 @@
+    private Map<String, Object> opts;
+    private ServerContext serverContext;
+
+    private IFailureDetectorPolicy failureDetectorPolicy;
```
This has a race condition. Ping counters can be over-written by late requests.

The comment says "We don't care if an increment gets clobbered by another increment", this is not true.
The behavior is non-deterministic, there is a race, so we can't assume that the maximum counter difference is 1.

zalokhan replied:
You are right but if there is a transient failure, this race condition does not cause an issue.
If the node has actually failed, the failure will certainly show up in a few pings.

If I correct this race by a blocking call, won't it will slow down the detection? (Consider round-robbin poll of 1000 nodes.)

Maithem replied:
Allowing a race condition for "performance" purposes doesn't sound right. I was also not implying that they should be blocking calls. 

Also, counting failed pings since the last successful ping might not be a good metrics, it would be able to discover a node failure, but not a fault. A slow host can miss the right pings such the the counter is always reset and never passes the threshold.

You don't have to change your policy/metric code in this patch, but this is something for you to think about.

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,236 @@
+        } while (corfuRuntime == null);
+
+
+        while (true) {
```
Why not use a ScheduledExecutorService ?

zalokhan replied:
I'll try putting the entire method as a ScheduledExecutorService rather than using a single thread and a while loop.

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,236 @@
+        if (corfuRuntime != null) {
+            corfuRuntime.isShutdown = true;
+            if (corfuRuntime.layout != null) {
+                corfuRuntime.layout.cancel(true);
```
Why do you care about cancelling the layout request, if you are shutting down?

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,236 @@
+
+                Thread.sleep(policyExecuteInterval);
+
+            } catch (InterruptedException err) {
```
Maybe just react to the interrupted exception instead of having to signal using the isShutdown variable?

zalokhan replied:
Ill change that.

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,236 @@
+
+        // Shutting the fault detector.
+        schedulerService.shutdownNow();
+        while (!schedulerService.isShutdown()) {
```
I don't think while loop is necessary, you can use awaitTermination

zalokhan replied:
Done

Maithem commented on src/main/java/org/corfudb/runtime/CorfuRuntime.java:
```diff
@@ -94,6 +94,13 @@
+    /**
+     * To prevent livelock condition in fetchLayout.
+     * Notifies that the runtime is no longer used
+     * and retries to fetch the layout can be stopped.
```
Maybe change layout.get to timeout instead of having to do this signaling ?

zalokhan replied:
Timeout will not be sufficient, there can be several layout servers in the list to which the runtime connects to. So there is no definite timeout value we can put there.
We can have retries instead. Lets say 3 complete rounds of all the layout servers.
This will delay the shutdown but I agree it will be a cleaner way to handle this.

Maithem replied:
I don't think this is necessary. It's also not a livelock condition.
The runtime calls an rpc on a single server every time it needs to fetch a layout.
The layout fetch is spun in a common pool, initiating a proper shutdown of that pool is more appropriate
than signaling using global variables.

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,236 @@
+     * @param layout        Current layout with failed nodes.
+     * @param failedServers Set of server nodes to be removed from the layout.
+     */
+    private void triggerHandlerDispatcher(Layout layout, Set<String> failedServers) {
```
Since this isnt an interface or abstract, then just add it when you want to implement it. There's no point of having this stub.

zalokhan replied:
Ok I will move this to my next PR.

dahliamalkhi commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,236 @@
+
+        while (true) {
+            try {
+                corfuRuntime.invalidateLayout();
```
Why are you invoking invalidateLayout on each polling interval? Unless I am missing something, this is wrong. It will unnecessarily cause an epoch-increase on each polling internal.

dahliamalkhi replied:
You are right, this does not invalidate the current layout (as the name might suggest), it invalidates previous ones. 

One question still remains: This does loop through all the layout servers. Do we want this communication round to occur upon each polling period?

zalokhan replied:
**This does loop through all the layout servers.**
The CorfuRuntime does loop through the layout servers until it connects to one. It then queries and returns the layout.
In normal conditions assuming all layout servers are operational, it will fetch the layout from the first one it queries.

**Do we want this communication round to occur upon each polling period?**
This is a tradeoff.
We can detect layout changes quickly with this.
Else we can batch polls into let's say batches of 5. Invalidate runtime once and check for failures once.

dahliamalkhi replied:
ok!

zalokhan commented:
I have made the fixes.
Pushed some of the fixes required in the periodicPollPolicy to the next PR.

dahliamalkhi commented on src/test/java/org/corfudb/infrastructure/ManagementServerTest.java:
```diff
@@ -0,0 +1,35 @@
+        managementServer.shutdown();
+        assertThat(managementServer.getFailureDetectorService().isShutdown());
+    }
+}
```
is it too early to test actual detection in this PR? that is, bring up a configuration, start failure detection, and then stop one component and verify it is detected faulty.

dahliamalkhi replied:
I was not asking to update layout; just to add a test that verifies that a failure is detected. Still too early?

dahliamalkhi replied:
ok, I guess the answer is yes, too early.. no rush then.

no2chem commented:
@Maithem - can you approve/request new changes so we can merge this PR?

medhavidhawan commented on src/main/java/org/corfudb/runtime/CorfuRuntime.java:
```diff
@@ -145,6 +144,23 @@ public CorfuRuntime() {
+        // Stopping async task from fetching layout.
+        isShutdown = true;
+        if (layout != null) {
+            try {
```
what is this code for ? There is no logging of exception ?

zalokhan replied:
Will log this

medhavidhawan commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -167,70 +168,41 @@ private void taskScheduler() {
+        failureDetectorPolicy.executePolicy(currentLayout, corfuRuntime);
+
+        // Get the server status from the policy and check for failures.
+        HashMap<String, Boolean> map = failureDetectorPolicy.getServerStatus();
```
This is more about code structure, what is the reasoning for having the server status as part of the policy ?

zalokhan replied:
getServerStatus gets the accumulated status of the nodes/servers since the last fetch.
So the policy keeps state of the status until another fetch occurs.

medhavidhawan commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -97,11 +100,9 @@ private Layout waitForBootstrap() {
             try {
```
With these while loops (forever kinds) please provide enough logging statements.

zalokhan replied:
I will add this.

slfritchie commented on src/main/java/org/corfudb/infrastructure/CorfuServer.java:
```diff
@@ -54,6 +54,9 @@
+    @Getter
+    private static ManagementServer managementServer;
+
     public static boolean serverRunning_p = false;
```
Hi, Medhavi.  That line was there before Zeeshan got to it.  It is my terrible style influence. I'm trying to correct my ways.  ^_^  But it will be a short while before cleanup is finished.  @zalokhan Feel free to change that name while you're there or leave it to me.

zalokhan replied:
Yeah I can refactor the name to serverRunning.
It will require a refactor in ICmdlet too so I'll add that change to this PR

zalokhan commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,236 @@
+                failureDetectorPolicy.executePolicy(currentLayout, corfuRuntime);
+
+                // Get the server status from the policy and check for failures.
+                HashMap<String, Boolean> map = failureDetectorPolicy.getServerStatus();
```
Ok, Ill add INFO logging here to show if failure detected.
If things are fine I'll add a DEBUG level else there will be too much output. One log statement every second.

zalokhan commented on src/main/java/org/corfudb/runtime/CorfuRuntime.java:
```diff
@@ -301,13 +308,19 @@ public void invalidateLayout() {
                         return l;
                     } catch (Exception e) {
                         log.warn("Tried to get layout from {} but failed with exception:", s, e);
+                        if (isShutdown) {
```
Fixed this in the latest commit.

zalokhan commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -97,11 +100,9 @@ private Layout waitForBootstrap() {
-                if (isShutdown) {
-                    return null;
-                }
+            } catch (InterruptedException ie) {
```
As per my understanding, the shutdownNow's best effort is sending an interrupt. If the thread fails to respond to this interrupt, it does not terminate.
Just to be safe, I can change the implementation to use a volatile flag.

zalokhan commented:
Added a couple of logging statements and removed total dependency on the interrupts to handle shutdown by adding a shutdown flag.

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,222 @@
+
+        this.failureDetectorPolicy = serverContext.getFailureDetectorPolicy();
+
+        this.failureDetectorService = Executors.newSingleThreadScheduledExecutor(
```
Should this take an executor or a single thread executor?

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,222 @@
+    /**
+     * Wait until we receive a message with the layout endpoint to bootstrap the management server.
+     */
+    private Layout waitForBootstrap() {
```
If you use a layout client this would be cleaner.

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,222 @@
+
+        // Get the server status from the policy and check for failures.
+        HashMap<String, Boolean> map = failureDetectorPolicy.getServerStatus();
+        if (!map.isEmpty()) {
```
Shouldn't this code belong in the policy itself? how does the management server know how to interpret these values? I feel that this logic shouldnt leak to the management server.

zalokhan replied:
Makes sense.
But I also feel the policy should be abstract and should not be able to trigger the failure handler directly. I can perhaps pass a model back from the policy which is passed to the failure handler and let the handler decide whether to take action or not. Would that be better?

Maithem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,222 @@
+        }
+
+        try {
+            failureDetectorService.awaitTermination(5, TimeUnit.SECONDS);
```
What is the significance of this 5 second wait?
Why not just do shutdownNow?

Maithem commented on src/test/java/org/corfudb/infrastructure/ManagementServerTest.java:
```diff
@@ -0,0 +1,35 @@
+        assertThat(!managementServer.getFailureDetectorService().isShutdown());
+        managementServer.shutdown();
+        assertThat(managementServer.getFailureDetectorService().isShutdown());
+    }
```
assertThat(managementServer.getFailureDetectorService().isShutdown()).isTrue()

Maithem commented on src/test/java/org/corfudb/infrastructure/ManagementServerTest.java:
```diff
@@ -0,0 +1,35 @@
+    public void checkFailureDetectorStatus() {
+        ManagementServer managementServer = getDefaultServer();
+
+        assertThat(!managementServer.getFailureDetectorService().isShutdown());
```
Better usage:
assertThat(managementServer.getFailureDetectorService().isShutdown()).isFalse()

Maithem commented on src/test/java/org/corfudb/infrastructure/ManagementServerTest.java:
```diff
@@ -0,0 +1,35 @@
+     * Testing the status of the failure detector and shutdown functionality.
+     */
+    @Test
+    public void checkFailureDetectorStatus() {
```
I don't see how this test is effective. The isShutdown flag is switched right after the method invocation, so essentially this test only verifies that the constructor and shutdown method have been called, nothing more. It also doesnt test the side effects of the shutdown and start operation.

zalokhan replied:
Yes it is meant to check the shutdown function only.
The pollingPolicy has its own test. I felt a more detailed test can be written once other modules are in place.
I did not want to expose unrelated data from the policy just for the test.

Maithem commented:
Also, please don't force push because that deletes our previous comments.

no2chem commented:
@Maithem I think the comments are still there, under show outdated

Maithem commented:
@no2chem For some, yes. I think github only tracks the diff snippits. For example, if you click on the first show outdated file, it wont find it.

zalokhan commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,222 @@
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Instantiates and performs failure detection and handling asynchronously.
```
I meant handling of detected failures.
I'll correct this if this is unclear.

zalokhan commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,222 @@
+ * Failure Detector:
+ * Executes detection policy (eg. PeriodicPollingPolicy).
+ * It then checks for status of the nodes. If the result map is not empty,
+ * there are failed nodes to be addressed. This then triggers the failure
```
Result map is just the result of the statuses of the servers received after polling.
You are right, the management server is a lot more than just failure detection.
Since I was explaining the failure detector, I have mentioned failure detector on top.

The failure detector of the management server is supposed to periodically execute the policy and depending on the status trigger the failure handler.

zalokhan commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,222 @@
+
+    @Override
+    public void reboot() {
+    }
```
There are other things to be handled as well apart from just the new policy, I have in mind.
So I thought of pushing the reboot and reset functions to the next PR.

zalokhan commented:
Resolved Management Server bootstrapping issue.
Now waits for bootstrapping through message handler and also persists the layout after every update in the server context datastore.

Now handling failures through message handlers than synchronous handling after polling.

no2chem commented on src/main/java/org/corfudb/cmdlets/corfu_layout.java:
```diff
@@ -153,6 +154,7 @@
             log.debug("Bootstrapping with layout {}", l);
             try {
                 if (router.getClient(LayoutClient.class).bootstrapLayout(l).get()) {
+                    router.getClient(ManagementClient.class).bootstrapManagement(l).get();
```
Shouldn't you be checking the return value of this bootstrap call? Also, I don't think this belongs in corfu_layout, which is supposed to be for layout servers. Please provide a clojure script which does this work.

no2chem commented on src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -0,0 +1,314 @@
+    }
+
+    @Override
+    public synchronized void handleMessage(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
```
Please change this to use the @ServerHandler annotation (see LogUnitServer for details)

no2chem commented on src/main/java/org/corfudb/protocols/wireprotocol/CorfuMsgType.java:
```diff
@@ -68,7 +67,11 @@
+
+    // Management Codes
+    MANAGEMENT_BOOTSTRAP(62, new TypeToken<CorfuPayloadMsg<ManagementBootstrapRequest>>(){}, ManagementServer.class, true),
+    FAILURE_DETECTED(63, new TypeToken<CorfuPayloadMsg<FailureDetectorMsg>>(){}, ManagementServer.class, true);
```
This should be MANAGEMENT_FAILURE_DETECTED.

no2chem commented on src/main/java/org/corfudb/protocols/wireprotocol/FailureDetectorMsg.java:
```diff
@@ -0,0 +1,26 @@
+@Data
+@AllArgsConstructor
+public class FailureDetectorMsg implements ICorfuPayload<FailureDetectorMsg> {
+    private Map nodes;
```
Please type (use the generic version) of Map.

no2chem commented on src/main/java/org/corfudb/protocols/wireprotocol/ManagementBootstrapRequest.java:
```diff
@@ -0,0 +1,25 @@
+ */
+@Data
+@AllArgsConstructor
+public class ManagementBootstrapRequest implements ICorfuPayload<ManagementBootstrapRequest> {
```
This class is unneeded. Can you just pass in CorfuMsg<Layout> to your handlers instead of creating a data type which just contains a layout?

zalokhan commented:
Refactored and simplified Management server related Corfu messages.
Modified to use @ServerHandler
Added clojure cmdlet for bootstrapping management server

no2chem commented on src/test/java/org/corfudb/runtime/view/ManagementViewTest.java:
```diff
@@ -0,0 +1,68 @@
+
+        // Wait in loop until failures have been reported.
+        while (!ManagementViewTest.failureDetectorMsgRecd)
+            Thread.sleep(500);
```
Can you add a timeout to this loop? I don't want the test to hang if it fails somehow. Probably use a semaphore with a timeout on the acquire.

zalokhan commented:
Added semaphore to fix test running indefinitely.
